bug(deno): End-to-end tests flake — serve/close_all teardown trips Deno leak sanitizer (analogue of #243)

## Summary

The CI **End-to-end tests** job ("Deno adapter + compliance") intermittently
fails on Deno's op/resource **leak sanitizer** at *teardown*. Every functional
subtest passes (`93 passed, 0 failed`), but the run is marked `FAILED` because
the sanitizer reports `Leaks detected` (or, less often,
`AssertionError: serve loop did not exit after close_all`) for one of the
teardown-sensitive regression tests.

This is the **Deno-side analogue of #243** (the Node `--test-force-exit`
teardown SIGABRT we fixed in #244). The Node teardown path is now hardened; the
Deno serve/close_all teardown path is not.

## Affected tests

Both live in `packages/iroh-http-deno/test/adapter.test.ts` (~line 91) and use
Deno's `sanitizeOps` / `sanitizeResources` leak detection:

- `serve — no pending ops remain after signal abort (regression #115)`
- `serve — close_all wakes pending polling loop (regression #155)`

A telltale of flakiness: it's a **different test each run** while all functional
assertions pass — characteristic of a teardown timing race, not a regression.

## Evidence (recent runs)

- Run `28179359603` (pre-release `main`): failed on #155 (`serve loop did not
  exit after close_all`), then on a rerun failed on #115 (`Leaks detected`),
  then went green on a 3rd rerun.
- Run `28180713817` (v0.5.1 release commit `6c9646d`): failed on #115
  (`Leaks detected`), 93/93 functional subtests passing.

## Not a publish blocker (but still noisy)

Publishing gates on the **Extended tests** workflow (a separate workflow), whose
Deno job passed cleanly for v0.5.1 — so 0.5.1 published correctly. This flake is
in the regular **CI** `End-to-end tests` job, which is not a publish gate.
Impact is therefore:

- Red CI on `main` and on release commits even when the build is good.
- Per-release "is this real?" triage and repeated reruns.

## Likely root cause

The serve `close_all` / signal-abort teardown path does not fully drain the
polling loop's pending ops/resources before the test boundary: a wakeup op, the
abort-signal listener, or a still-draining response pipe survives past the
sanitizer snapshot. The polling loop is woken to exit, but the op/resource it
held isn't always settled synchronously with `close_all` resolving.

## Proposed fix

1. Ensure `close_all` (and the signal-abort path) **awaits full drain** of the
   serve polling loop — every pending op/timer/resource released — before it
   resolves, so the sanitizer sees a clean slate.
2. If a late endpoint-gone settlement is the culprit, swallow it on the Deno
   adapter teardown path the same way the other endpoint-gone classes
   (#119/#122/#161) are handled.
3. Keep the regression assertions intact so a genuine leak still fails.

## Acceptance criteria

- The two regression tests pass reliably under
  `deno test --allow-... packages/iroh-http-deno/test/adapter.test.ts` with op
  and resource sanitizers enabled, across repeated/looped runs.
- A genuine op/resource leak still trips the sanitizer (guard is scoped to the
  benign teardown settlement only).
- CI `End-to-end tests` (Deno) is green on a clean build without reruns.

--- a/packages/iroh-http-deno/src/adapter.ts
+++ b/packages/iroh-http-deno/src/adapter.ts
@@ -248,6 +248,13 @@ function createYieldFn(): {
 // the -1 shutdown sentinel without a timing-dependent delay.
 const serveCancellers = new Map<number, () => void>();
 
+// #245: The `stopServe` FFI call is a non-blocking op (`iroh_http_call`).
+// If it is left floating it can still be in flight when a test's `sanitizeOps`
+// snapshot is taken, which Deno's leak sanitizer reports as a leaked op. This
+// map hands the in-flight promise to the serve loop so it can drain the op
+// before resolving `finished`. Entries self-clean once the op settles.
+const serveStopOps = new Map<number, Promise<unknown>>();
+
 // ── JSON dispatch helper ──────────────────────────────────────────────────────
 
 const enc = new TextEncoder();
@@ -783,6 +790,10 @@ export const rawServe: RawServeFn = (
         } finally {
           serveCancellers.delete(endpointHandle);
           yielder.close();
+          // #245: Drain the in-flight stopServe op (if any) before resolving
+          // so no non-blocking FFI op survives past `handle.finished`.
+          const stopOp = serveStopOps.get(endpointHandle);
+          if (stopOp) await stopOp;
         }
       })();
     },
@@ -806,6 +817,11 @@ export function makeAllocBodyWriter(endpointHandle: number): AllocBodyWriterFn {
 // Sends QUIC CONNECTION_CLOSE frames so peers observe a clean disconnect.
 function _closeAllSync(): void {
   lib.symbols.iroh_http_close_all();
+  // #245: close_all removes the request queues, but a serve loop parked in
+  // `await yielder.yield()` only re-polls when its MessageChannel turn lands.
+  // Wake every loop explicitly so it sees the -1 sentinel promptly instead of
+  // relying on MessageChannel timing under load.
+  for (const cancel of serveCancellers.values()) cancel();
 }
 Deno.addSignalListener("SIGTERM", _closeAllSync);
 // SIGINT is registered only on platforms that support it (non-Windows).
@@ -909,9 +925,15 @@ export function stopServe(handle: number): void {
   // #115: Cancel the yield immediately so the polling loop re-polls and
   // sees the -1 shutdown sentinel without waiting for the MessageChannel.
   serveCancellers.get(handle)?.();
-  call<Record<never, never>>("stopServe", { endpointHandle: handle }).catch(
-    () => {},
-  );
+  // #245: Register the in-flight op so the serve loop can await it before
+  // resolving `finished`; otherwise this non-blocking FFI op can outlive the
+  // test boundary and trip Deno's `sanitizeOps` leak detector under load.
+  const op = call<Record<never, never>>("stopServe", { endpointHandle: handle })
+    .catch(() => {});
+  serveStopOps.set(handle, op);
+  void op.finally(() => {
+    if (serveStopOps.get(handle) === op) serveStopOps.delete(handle);
+  });
 }
 
 /** Resolves when the endpoint has been fully closed (explicit or native). */
@@ -1313,11 +1335,9 @@ export class DenoAdapter extends IrohAdapter {
   }
 
   stopServe(handle: number): void {
-    // #115: Cancel the yield immediately so the polling loop re-polls.
-    serveCancellers.get(handle)?.();
-    call<Record<never, never>>("stopServe", { endpointHandle: handle }).catch(
-      () => {},
-    );
+    // Delegate to the module-level stopServe so the in-flight op is tracked
+    // and drained by the serve loop (#115/#245).
+    stopServe(handle);
   }
 
   waitEndpointClosed(handle: number): Promise<void> {

